libs/MotionKit/source/MotionKit.cpp

```diff
@@ -0,0 +1,113 @@
+// Leka - LekaOS
+// Copyright 2022 APF France handicap
+// SPDX-License-Identifier: Apache-2.0
+
+#include "MotionKit.h"
+
+#include "MathUtils.h"
+#include "ThisThread.h"
+
+using namespace leka;
+using namespace std::chrono_literals;
+
+void MotionKit::init()
+{
+	_event_loop.registerCallback([this] { run(); });
+}
+
+void MotionKit::stop()
+{
+	_left_motor.stop();
+	_right_motor.stop();
+	_is_not_delay_over		  = false;
+	_target_not_reached		  = false;
+	_rotate_x_turns_requested = false;
+	_event_loop.stop();
+}
+
+void MotionKit::controlledSpin(uint8_t number_of_turn, Rotation direction)
+{
+	stop();
+
+	_imu_kit.start();
+	_imu_kit.reset();
+	_target_not_reached		  = true;
+	_is_not_delay_over		  = false;
+	_rotate_x_turns_requested = true;
+
+	_turn_to_execute = number_of_turn;
+	_turn_executed	 = 0;
+
+	_left_motor.spin(direction, kPwmMax);
+	_right_motor.spin(direction, kPwmMax);
+
+	_event_loop.start();
+}
+
+void MotionKit::responsive(std::chrono::seconds duration)
+{
+	stop();
+
+	_imu_kit.start();
+	_imu_kit.reset();
+	_target_not_reached		  = true;
+	_is_not_delay_over		  = true;
+	_rotate_x_turns_requested = false;
+
+	auto on_timeout = [this] { _is_not_delay_over = false; };
+	_timeout.onTimeout(on_timeout);
+	_timeout.start(duration);
+
+	_event_loop.start();
+}
+
+// LCOV_EXCL_START - Dynamic behavior, involving motors and time.
+void MotionKit::run()
+{
+	auto last_yaw = kReferenceAngle;
+	while (_rotate_x_turns_requested) {
+		auto [current_pitch, current_roll, current_yaw] = _imu_kit.getAngles();
+
+		if (std::abs(last_yaw - current_yaw) >= 300.F) {
+			++_turn_executed;
+		}
+		rtos::ThisThread::sleep_for(70ms);
+
+		last_yaw = current_yaw;
+		if (_turn_executed == _turn_to_execute) {
+			_rotate_x_turns_requested = false;
+			_turn_executed			  = 0;
+			_turn_to_execute		  = 0;
+		}
+	}
```

Comment:
A suggestion to simplify what happens inside the while loop

```suggestion
	auto last_yaw = kReferenceAngle;

	auto turn_executed				  = 0;
	auto must_rotate				  = [&] { return _rotate_x_turns_requested && turn_executed != _turn_to_execute; };
	auto check_complete_turn_executed = [&](auto last_yaw, auto current_yaw) {
		if (std::abs(last_yaw - current_yaw) >= 300.F) ++turn_executed;
	};

	while (must_rotate()) {
		auto [current_pitch, current_roll, current_yaw] = _imu_kit.getAngles();

		check_complete_turn_executed(last_yaw, current_yaw);
		rtos::ThisThread::sleep_for(70ms);

		last_yaw = current_yaw;
	}
	_rotate_x_turns_requested = false;
	_turn_to_execute		  = 0;
```

Reply:
Move out of a loop what happens at the exit of it, not inside as 

``` cpp
		if (_turn_executed == _turn_to_execute) {
			_rotate_x_turns_requested = false;
			_turn_executed			  = 0;
			_turn_to_execute		  = 0;
		}
```

suggests